api/protocol/lorawan/validation.go

```diff
@@ -1,121 +1,167 @@
 	}
 	if m.NwkSKey != nil && m.NwkSKey.IsEmpty() {
-		return false
+		return errors.NewErrInvalidArgument("NwkSKey", "can not be empty")
 	}
-	return true
+	return nil
 }
 
 // Validate implements the api.Validator interface
-func (m *Message) Validate() bool {
+func (m *Message) Validate() error {
 	if m.Major != Major_LORAWAN_R1 {
-		return false
+		return errors.NewErrInvalidArgument("Major", "invalid value " + Major_LORAWAN_R1.String())
 	}
 	switch m.MType {
 	case MType_JOIN_REQUEST:
-		return m.GetJoinRequestPayload() != nil && m.GetJoinRequestPayload().Validate()
+		if m.GetJoinRequestPayload() == nil {
+			return errors.NewErrInvalidArgument("JoinRequestPayload", "can not be empty")
+		}
+		if err := m.GetJoinRequestPayload().Validate(); err != nil {
+			return errors.NewErrInvalidArgument("JoinRequestPayload", err.Error())
+		}
 	case MType_JOIN_ACCEPT:
-		return m.GetJoinAcceptPayload() != nil && m.GetJoinAcceptPayload().Validate()
+		if m.GetJoinAcceptPayload() == nil {
+			return errors.NewErrInvalidArgument("JoinAcceptPayload", "can not be empty")
+		}
+		if err := m.GetJoinAcceptPayload().Validate(); err != nil {
+			return errors.NewErrInvalidArgument("JoinAcceptPayload", err.Error())
+		}
 	case MType_UNCONFIRMED_UP, MType_UNCONFIRMED_DOWN, MType_CONFIRMED_UP, MType_CONFIRMED_DOWN:
-		return m.GetMacPayload() != nil && m.GetMacPayload().Validate()
+		if m.GetMacPayload() == nil {
+			return errors.NewErrInvalidArgument("MacPayload", "can not be empty")
+		}
+		if err := m.GetMacPayload().Validate(); err != nil {
+			return errors.NewErrInvalidArgument("MacPayload", err.Error())
+		}
+	default:
+		return errors.NewErrInvalidArgument("MType", "unknown type " + m.MType.String())
 	}
-	return false
+
+	return nil
 }
 
 // Validate implements the api.Validator interface
-func (m *JoinRequestPayload) Validate() bool {
-	return len(m.AppEui) == 8 && len(m.DevEui) == 8 && len(m.DevNonce) == 2
+func (m *JoinRequestPayload) Validate() error {
+	if len(m.AppEui) != 8 {
+		return errors.NewErrInvalidArgument("AppEui", "length must be 8")
+	}
+	if len(m.DevEui) != 8 {
+		return errors.NewErrInvalidArgument("DevEui", "length must be 8")
+	}
+	if len(m.DevNonce) != 2 {
+		return errors.NewErrInvalidArgument("DevNonce", "length must be 2")
+	}
+
+	return nil
 }
 
 // Validate implements the api.Validator interface
-func (m *JoinAcceptPayload) Validate() bool {
+func (m *JoinAcceptPayload) Validate() error {
 	if len(m.Encrypted) != 0 {
-		return true
+		return nil
 	}
+
 	if m.CfList != nil && len(m.CfList.Freq) != 5 {
-		return false
+		return errors.NewErrInvalidArgument("CfList.Freq", "length must be 5");
+	}
+
+	if len(m.DevAddr) != 4 {
+		return errors.NewErrInvalidArgument("DevAddr", "length must be 4")
 	}
-	return len(m.DevAddr) == 4 && len(m.AppNonce) == 3 && len(m.NetId) == 3
+	if len(m.AppNonce) != 3 {
+		return errors.NewErrInvalidArgument("AppNonce", "length must be 3")
+	}
+	if len(m.NetId) != 3 {
```

Comment:
Change this to use `Empty()` method (create if doesn't exist; length of a `types.NetID` is always 3)